tests/multigrid/test_generalised_hierarchy.py

```diff
@@ -0,0 +1,175 @@
+from firedrake import *
+from firedrake.mg import *
+from firedrake.mg.utils import *
+
+import numpy as np
+
+import pytest
+
+
+def test_generalised_mesh_hierarchy():
+
+    mesh = UnitSquareMesh(5, 5)
+    L = 4
+    GMH = GeneralisedMeshHierarchy(mesh, L)
+
+    assert len(GMH) == L + 1
+
+    assert GMH.M == 2
+
+    a = 0
+    try:
+        GMH = GeneralisedMeshHierarchy(mesh, L, 3)
+
+    except ValueError:
+        a = 1
+
+    assert a == 1
+
+    L = 3
+    M = 4
+    GMH = GeneralisedMeshHierarchy(mesh, L, M)
+
+    assert len(GMH) == L + 1
+
+
+def test_attribute():
+
+    mesh = UnitSquareMesh(5, 5)
+    L = 3
+    M = 4
+    GMH = GeneralisedMeshHierarchy(mesh, L, M)
+
+    assert hasattr(GMH, '_full_hierarchy') == 1
+
+    GFSH = GeneralisedFunctionSpaceHierarchy(GMH, 'DG', 0)
+
+    assert hasattr(GFSH, '_full_hierarchy') == 1
+
+    F = Function(GFSH[0])
+
+    assert isinstance(get_level(F.function_space())[0], FunctionSpaceHierarchy) == 1
+
+    assert hasattr(get_level(F.function_space())[0], '_full_hierarchy')
+
+    assert hasattr(get_level(F.function_space())[0], '_skip')
+
+
+def test_warning_non_generalised_mesh_hierarchy():
+
+    mesh = UnitSquareMesh(5, 5)
+    L = 3
+    MH = MeshHierarchy(mesh, L)
+
+    a = 0
+
+    try:
+        GeneralisedFunctionSpaceHierarchy(MH, 'DG', 0)
+
+    except AttributeError:
+        a = 1
+
+    assert a == 1
+
+
+def test_get_level():
+
+    mesh = UnitSquareMesh(5, 5)
+    L = 3
+    M = 4
+    GMH = GeneralisedMeshHierarchy(mesh, L, M)
+
+    FSH = GeneralisedFunctionSpaceHierarchy(GMH, 'DG', 0)
+
+    assert len(FSH) == L + 1
+
+    lev = []
+    for i in range(L + 1):
+        F = Function(FSH[i])
+        lev.append(get_level(F.function_space())[1])
+
+    assert np.all(lev == np.linspace(0, L, L + 1)) == 1
+
+
+def test_get_full_level():
+
+    mesh = UnitSquareMesh(5, 5)
+    L = 3
+    M = 4
+    GMH = GeneralisedMeshHierarchy(mesh, L, M)
+
+    FSH = GeneralisedFunctionSpaceHierarchy(GMH, 'DG', 0)
+
+    assert len(FSH) == L + 1
+
+    lev = []
+    for i in range((L * (M / 2)) + 1):
+        F = Function(FSH._full_hierarchy[i])
+        lev.append(get_level(F.function_space())[1])
+
+    assert np.all(lev == np.linspace(0, (L * (M / 2)), (L * (M / 2)) + 1)) == 1
```

Comment:
Actually, don't both collecting the levels, just `assert get_level(...)[1] == i` in the loop.